### ISSUE TITLE
Feature request: propagate return types of do-like expressions

### ISSUE THREAD

**Issue by yuhan0** [clj-kondo/clj-kondo#2336]
- [x] I have read the [Clojure etiquette](https://clojure.org/community/etiquette) and will respect it when communicating on this platform.

**Is your feature request related to a problem? Please describe.**

The :type-mismatch linter does not catch errors like
```clj
(inc (do "not a number"))

(cycle (doto 10 (repeat)))
```

**Describe the solution you'd like**
Type annotations could be used to propagate the return type.

**Additional context**

Apart from `do` and `doto` which were implemented in #2334 (closed PR), a quick search through clojure.core for `& body` finds these similar do-like macros where the return type could be trivially analysed as the 'last' body expression.  

### not yet annotated
- locking
- when-some
- binding
- with-bindings
- sync
- io!
- with-open
- with-local-vars
- loop
- when-first
- with-in-str
- dosync
- with-precision
- with-loading-context
- letfn
- with-redefs

### already annotated
- when
- when-not
- let
- when-let

**Comment by borkdude** [comment]
I think the scope of this issue is too wide. I'll limit it to just `do`. Others can be discussed in other issues as the solution for each may be different and more problematic.

**Comment by borkdude** [comment]
Submitted a PR here which just fixes `do`:

https://github.com/clj-kondo/clj-kondo/pull/2337

and made you co-author.

**Comment by borkdude** [comment]
Although this issue is closed, PR welcome on a case by case basis for other stuff

**Comment by yuhan0** [comment]
Great, thanks so much! 
Just out of curiosity, I got quite stuck yesterday trying to figure out the reasoning for the `add-new-arg-types?=false` argument to `analyze-children`, besides the feature not working otherwise..

In both true/false cases the local expressions appear to be identical `(:arg-types ctx)` ~= `(atom[])`, so the difference lies in the reference to the specific atom being mutated elsewhere?

I'd appreciate if you could give quick pointers to where / how that action-at-a-distance is happening in this specific case  - maybe if it helps me build a mental model of the codebase I could turn that into a couple of code comments or docstrings in analyzer.clj to help make it friendlier for future contributors :)

**Comment by borkdude** [comment]
@yuhan0 The gist is that when calling `analyze-children` with `true` (which is the default), it will insert a new atom in which types of child nodes are collected. Doing so with `false` will collect them for `do` which is the intent

**Comment by borkdude** [comment]
I have to admit that I find this difficult to understand myself every time I look at it after a while.
